### DISCUSSION TITLE
Creating records via API Route when using Pocketbase as Framework

### DISCUSSION THREAD

**Discussion by Y4shin** [pocketbase/pocketbase#734] [Q&A] [answered]
I am currently implementing a API route while using PocketBase as a Framework that is, among other things, inserting Data into the database. However I have not been able to figure out, how to properly do this.

I have used the `app.DB().Model(&object).Insert()` function until now, however I have the Problem, that I do not know how to properly create an id for the new record. I can just create a random 15 character String, but I do not think, that that is the proper way of doing things, especially as it does not guarantee that there is no colision with an existing record.

What would be the proper way of creating new Records in an API route.

Here would be an example of a handler function how I have done it right now:
```go
func random15CharString() string {
    //...
}

type RecordType struct {
    Id        string
    SomeValue string
}

func handleRoute(c echo.Context) error {
    //...
    newRecord := RecordType{
        Id:        random15CharString(),
        SomeValue: "value",
    }
    app.DB().Model(&newRecord).Insert()
    //...
}
```

**Comment by ganigeorgiev** [comment, marked as answer]
You rarely will have a need to use directly the `DB()` instance. In PocketBase there are 2 subpackages to help working with persisting data:

- **daos** - serves as repository and service layer in one, aka. can be used to query and persist your models. The daos methods usually don't perform data validations and directly translates your model structure into the appropriate db statements. You can find godoc for all available `daos` public definitions in https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.7.9/daos.

- **forms** - are usually "request forms", aka. they are safe to be loaded with external user data and are able to perform validations (eg. required checks). Most forms have a `Submit` method which first validate all the user data and if everything is ok, will call the related daos method for persistence. You can find godoc for all available `forms` public definitions in https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.7.9/forms

To create a new record programmatically you can create a new `Record` model and directly persist it with `app.Dao().SaveRecord(record)` (you can find godoc for all models in https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.7.9/models).

```go
collection, err := app.Dao().FindCollectionByNameOrId("test")
if err != nil {
	return err
}

// create a new record instance
// or fetch existing, eg. app.Dao().FindFirstRecordByData(collection, "yourField", "yourValue")
record := models.NewRecord(collection)
record.SetDataValue("yourField1", "yourValue1")
record.SetDataValue("yourField2", "yourValue2")

// upsert
if err := app.Dao().SaveRecord(record); err != nil {
	return err
}
```

You don't need to specify an id, it will be auto generated before insert.

---

If you want to have a typed structure for your collection records, you can create your own model and use the `app.Dao().save(model)` method to persist it. The only requirement is to implement the `models.Model` interface. You can do that easily by embedding `models.BaseModel` and specify only a `TableName()` method that return your collection name:

```go
var _ models.Model = (*YourCustomRecord)(nil)

type YourCustomRecord struct {
    models.BaseModel

    SomeField string `db:"someField" json:"someField"`
}

func (m *YourCustomRecord) TableName() string {
    return "example" // the name of your collection
}
```
And then you'll be able to save it like this:

```go
yourCustomRecord := &YourCustomRecord{}
yourCustomRecord.SomeField = "test"
err := app.Dao().Save(yourCustomRecord)
// or to delete
err := app.Dao().Delete(yourCustomRecord)
```
You can even create your own dao helpers similar to the core models. For a simple example, you could check the Admin dao helpers - https://github.com/pocketbase/pocketbase/blob/master/daos/admin.go

**Reply by ganigeorgiev** [reply]
_New documentation with more in depth guides and examples will be available very soon._

**Reply by Y4shin** [reply]
Okay, that is actually really cool. Thanks for the help. Also thanks for the *really* quick answer just like with my last question. I really appreciate that.
